File: backend/src/main/java/org/naukma/raft/entity/Reminder.java
package org.naukma.raft.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

/**
 * Represents a reminder created by a user.
 *
 * A reminder can be connected either to a task or to an event.
 * When its reminder time comes, the system can create a notification
 * and mark the reminder as sent.
 */
@Entity
@Table(name = "reminders")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Reminder {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "task_id")
    private Task task;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "event_id")
    private Event event;

    @Column(name = "reminder_time", nullable = false)
    private LocalDateTime reminderTime;

    @Column(name = "is_sent", nullable = false)
    private boolean isSent = false;
}

File: frontend/src/types/reminder.ts
/** A reminder for a task or event (exactly one of `taskId`/`eventId` is set). */
export interface Reminder {
    id: string;
    taskId?: string;
    eventId?: string;
    /** When to fire, as a local `YYYY-MM-DDTHH:mm` string. */
    reminderTime: string;
    /** True once the scheduler has sent the notification. */
    sent: boolean;
}

/** Payload to create a reminder (attach to a task or an event). */
export interface CreateReminderInput {
    taskId?: string;
    eventId?: string;
    reminderTime: string;
}

/** Payload to update a reminder (e.g. reschedule its time). */
export interface UpdateReminderInput {
    taskId?: string;
    eventId?: string;
    reminderTime?: string;
}